### PR TITLE
[ParticleMechanics] Adding unit tests and some cleanups

### PR DESCRIPTION
**Pull request by bodhinandach**
@antonialarese @philbucher @loumalouomega 
This PR is to add mainly the following enhancements to ParticleMechanicsApplication:
- Adding (**finally**) the first batch of tests to ParticleMechanicsApplication. I know that there should be more tests done to check all the functionalities. However, I don't want to make the PR to be too large and therefore will add more tests in the near future along the development of the codes.
- There is bug that I have fixed from my last PR of refactoring #3115.
- Some doxygen documentation in utilities and constitutive laws (will add more documentation in another part of the codes in the near future).
- Some cleanups of unused functions.
- I added also a utility to output json and check from json as the one in the core is calculating the elemental value from gauss points, while my particle element is storing the value in the element.

Please take some times to review this PR.

### CODE REVIEW

File: applications/ParticleMechanicsApplication/custom_utilities/mpm_stress_principal_invariants_utility.h [hunk 1]
```diff
@@ -41,7 +41,16 @@ namespace Kratos
 
             typedef unsigned int SizeType;
 
-            // Sort Principal Stresses, Strains, and Directions to magnitude order
+            // Tolerance
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_utilities/mpm_stress_principal_invariants_utility.h:44]: Most of this utilities are duplicated respect constitutive_law_utilities on StructuralMechanicsApplication

File: applications/ParticleMechanicsApplication/custom_utilities/mpm_stress_principal_invariants_utility.h [hunk 2]
```diff
@@ -257,45 +345,167 @@ namespace Kratos
...
-                  r2C3 *= - std::sqrt(3.0) / (2.0 * std::cos(3*lode_angle) * std::pow(j_2, 1.5));
+            /**
+             * @brief Transform Stress or Principal stress tensor to Vector form (fully assuming 3D space).
+             * @param[in] rMatrix Matrix tensor to be transformed.
+             * @param[in] rSize Matrix size, 3 for only principal stresses, 6 for normal stresses.
+             * @return transformed vector.
+             */
+            static Vector PrincipalMatrixtoVector(const Matrix& rMatrix, const unsigned int& rSize)
+            {
+                  Vector vector = ZeroVector(3);
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_utilities/mpm_stress_principal_invariants_utility.h:498]: Check constitutive_law_utilities

File: applications/ParticleMechanicsApplication/custom_utilities/particle_mechanics_math_utilities.h [hunk 3]
```diff
@@ -62,21 +62,137 @@ class ParticleMechanicsMathUtilities
...
-     * @param zero number treated as zero
-     * @return Vector of eigenvalues
-     * WARNING only valid for 2*2 and 3*3 Matrices yet
+     * @brief Calculates the QR Factorization of given square matrix A=QR.
+     *        The Factorization is performed using the householder algorithm.
+     * @param[in]  A The given square matrix the factorization is to be calculated.
+     * @param[out] Q The result matrix Q.
+     * @param[out] R The result matrix R.
      */
+    static inline void QRFactorization(const MatrixType& A, MatrixType& Q, MatrixType& R)
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_utilities/particle_mechanics_math_utilities.h:71]: Is this in MathUtils?, if not this is quite general and should go there

File: applications/ParticleMechanicsApplication/custom_utilities/particle_mechanics_math_utilities.h [hunk 4]
```diff
@@ -62,21 +62,137 @@ class ParticleMechanicsMathUtilities
...
+     * @param[in] A the given square matrix the eigenvalues are to be calculated.
+     * @param[in] rTolerance convergence criteria.
+     * @param[in] rZeroTolerance number treated as zero.
+     * @return Vector of eigenvalues.
+     * @warning Only valid for 2*2 and 3*3 Matrices yet.
+     */
+    static inline Vector EigenValues(const Matrix& A, const double rTolerance = 1e-9, const double rZeroTolerance = 1e-9)
     {
-        int dim= A.size1();
+        unsigned int dim= A.size1();
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_utilities/particle_mechanics_math_utilities.h:195]: This is on MathUtils

File: applications/ParticleMechanicsApplication/custom_utilities/particle_mechanics_math_utilities.h [hunk 5]
```diff
@@ -227,157 +342,34 @@ class ParticleMechanicsMathUtilities
...
-				    MatrixType& vectors,
-				    VectorType& lambda,
-				    double zero_tolerance =1e-9,
-				    int max_iterations = 10)
+				    MatrixType& rEigenVectors,
+				    VectorType& rEigenValues,
+				    const double rZeroTolerance = 1e-9,
+				    const unsigned int rMaxIteration = 10)
     {
+        if(A.size1()!=3 || A.size2()!=3)
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_utilities/particle_mechanics_math_utilities.h:360]: MathUtils!!!

File: applications/ParticleMechanicsApplication/custom_utilities/mpm_stress_principal_invariants_utility.h (hunk 1 above)
Comment (bodhinandach) [reply applications/ParticleMechanicsApplication/custom_utilities/mpm_stress_principal_invariants_utility.h:44]: Agree with you, but I am having a bit more specialized for soil with first and second derivatives.

**Comment by loumalouomega** [comment]
@bodhinandach even if your utilities are for soil, there are common operations and in any case this are still CL operations, that can be used by other people maybe not working in soils, but metals or other materials. Please refactor that

**Comment by bodhinandach** [comment]
@loumalouomega Yes, for sure. That's why I made it static so that other people might use it as well. I am not sure what I have to refactor...

**Comment by loumalouomega** [comment]
I mean to put it in the constitutive_law_utilities.h/cpp from StructuralMechanicsApplication, so everyone can use it (without the need of doing abstract includes). Besides, I am currently doing a review, but ther are so many changes, it will take me a while.

File: applications/ParticleMechanicsApplication/custom_utilities/particle_mechanics_math_utilities.h [hunk 6]
```diff
@@ -968,84 +632,133 @@ class ParticleMechanicsMathUtilities
...
 
+
+    /**
+    * @brief Builds the Inverse of Matrix input
+    * @param[in] rInput the given Matrix
+    * @param[out] rInverse inverse of the given Matrix
+    */
+    static int InvertMatrix( const MatrixType& rInput, MatrixType& rInverse )
+    {
+    #ifdef KRATOS_USE_AMATRIX   // This macro definition is for the migration period and to be removed afterward please do not use it
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_utilities/particle_mechanics_math_utilities.h]: This is on MathUtils

File: applications/ParticleMechanicsApplication/particle_mechanics_application.cpp [hunk 7]
```diff
@@ -183,9 +183,9 @@ namespace Kratos
 
         // Registering Constitutive Laws
         // CL: Linear Elastic laws
-        KRATOS_REGISTER_CONSTITUTIVE_LAW("LinearElastic3DLaw", mHyperElastic3DLaw);
-        KRATOS_REGISTER_CONSTITUTIVE_LAW("LinearElasticPlaneStress2DLaw", mHyperElasticPlaneStrain2DLaw);
-        KRATOS_REGISTER_CONSTITUTIVE_LAW("LinearElasticPlaneStrain2DLaw", mHyperElasticPlaneStrainUP2DLaw);
+        KRATOS_REGISTER_CONSTITUTIVE_LAW("LinearElastic3DLaw", mLinearElastic3DLaw);
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/particle_mechanics_application.cpp:186]: Which are the problems with the ones already in the StructuralMechanicsApplication

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Please refactor this. Move all the CL utilities that can be used for general pourpose to the constitutive_law_utilities in the StructuralMechanicsApplication. Try ti use the already implemented methods there.

Reduce code duplication with the MathUtils, if you have something of general interest on your math_utils please move it to the core. (add the corresponding tests)

**Comment by bodhinandach** [comment]
@loumalouomega Thanks for the comments. I am taking a closer look into it tomorrow.

**Review by philbucher — DISMISSED** [review]
@bodhinandach thanks for cleaning up and adding the tests

However ... this is a HUGE PR, including refactoring, cleanup, tests etc ...
=> I would really appreciate it if you could split it up into smaller PRs (you described it nicely in the first post, I would make a separate PR for each of the points)
This will make it a lot easier and faster for us to review

@loumalouomega it is true that some stuff is duplicated from stuff in StructuralMechanics. For now I would leave this as is, @bodhinandach just dropped other dependencies, I think we should give this app a bit of time to stabilize and consolidate until we start unifying things
Also bcs you and others are currently working a lot in the CLaws in StructuralMechanics

**Comment by loumalouomega** [comment]
@philbucher I already agreed with @bodhinandach in private to focus ily in the mat_utils duplications and postpone the CL utilities refactoring.

**Comment by philbucher** [comment]
@loumalouomega perfect!

**Comment by philbucher** [comment]
@bodhinandach also take a look at the codacy suggestions, they are relevant

**Comment by bodhinandach** [comment]
@philbucher great! I will fix some stuff tomorrow also considering codacy's suggestions. 
Sorry to make it too large. While making the tests, I started to fix things that were not correct and did some cleanups on top of that. But the main focus will be on the tests.

**Review by loumalouomega — APPROVED** [review]
Ok

**Comment by philbucher** [comment]
@bodhinandach in the future please do not dismiss reviews unless the person that requested the changes doesn't reply for weeks

please ping the person first

**Comment by bodhinandach** [comment]
@philbucher I truly apologize for that! :disappointed:  I will be careful next time. Many thanks in advance.

**Comment by philbucher** [comment]
@bodhinandach no worries ;)

**Pull request merged**